feat(lens): Add user-configurable sort control for heatmap axes

graphaelli opened the pull request:
Add sort order controls for X and Y axes in heatmap visualizations, allowing users to override auto-derived sort predicates with manual selection.

Features:
- Add sort order dropdowns in Vertical axis and Horizontal axis accordions
- Automatically detect numeric vs alphabetical data types and apply appropriate predicates (numAsc/numDesc for numbers, alphaAsc/alphaDesc for strings)
- Support for Auto (uses existing auto-detection), Ascending, Descending, and Original order (dataIndex) options
- Maintain backward compatibility: undefined predicates fall back to auto-detection

Implementation details:
- Added xSortPredicate and ySortPredicate optional fields to HeatmapGridConfig
- Updated heatmap_grid expression function to accept sort predicate arguments
- Enhanced UI components to detect column types from frame.activeData
- Updated heatmap_component to use user-provided predicates when available
- Added comprehensive unit tests for sort predicate functionality

Files changed:
- expression_functions.ts: Added type definitions for sort predicates
- heatmap_grid.ts: Added expression function arguments for sort predicates
- heatmap_component.tsx: Updated to use user-provided sort predicates
- style_settings.tsx: Added UI controls with automatic type detection
- visualization.tsx: Pass sort predicates through to expression
- style_settings.test.tsx: Added tests for sort control functionality

_Developed with Cursor (auto model)_

## Release Note

Sort order controls added for heatmap visualization axes.

## Summary

Given logs with http.server.request.duration in ms:

```
FROM logs-default-heatmap
| WHERE @timestamp >= ?_tstart AND @timestamp < ?_tend
| STATS count = COUNT(*) BY latency = BUCKET(http.server.request.duration, 5), time = TBUCKET(1 minute)
| SORT time
```

Before:

There were no sort order options - behavior retained with "Auto":

<img width="1887" height="701" alt="image" src="https://github.com/user-attachments/assets/8af44510-8626-4159-8500-3cab438b05e9" />

After:

<img width="1876" height="705" alt="image" src="https://github.com/user-attachments/assets/fdc47469-adaf-4e14-9b5d-1b450424b014" />

These settings also apply to non-esql lens heatmap visualizations, though the situation there is a bit different as empty buckets are displayed by default and the y-axis is sorted ascending by default:

<img width="1458" height="646" alt="image" src="https://github.com/user-attachments/assets/7816d0ea-5061-452a-85c2-a9a743b1bb54" />

This pr enables change the sort options for non-esql as well but it doesn't look much better.  A change for improving that will follow separately.


fix https://github.com/elastic/kibana/issues/241507
fix https://github.com/elastic/kibana/issues/239736

Commit:
feat(lens): Add user-configurable sort control for heatmap axes

Add sort order controls for X and Y axes in heatmap visualizations, allowing
users to override auto-derived sort predicates with manual selection.

Features:
- Add sort order dropdowns in Vertical axis and Horizontal axis accordions
- Automatically detect numeric vs alphabetical data types and apply appropriate
  predicates (numAsc/numDesc for numbers, alphaAsc/alphaDesc for strings)
- Support for Auto (uses existing auto-detection), Ascending, Descending, and
  Original order (dataIndex) options
- Maintain backward compatibility: undefined predicates fall back to auto-detection

Implementation details:
- Added xSortPredicate and ySortPredicate optional fields to HeatmapGridConfig
- Updated heatmap_grid expression function to accept sort predicate arguments
- Enhanced UI components to detect column types from frame.activeData
- Updated heatmap_component to use user-provided predicates when available
- Added comprehensive unit tests for sort predicate functionality

Files changed:
- expression_functions.ts: Added type definitions for sort predicates
- heatmap_grid.ts: Added expression function arguments for sort predicates
- heatmap_component.tsx: Updated to use user-provided sort predicates
- style_settings.tsx: Added UI controls with automatic type detection
- visualization.tsx: Pass sort predicates through to expression
- style_settings.test.tsx: Added tests for sort control functionality

graphaelli commented:
@elasticmachine merge upstream

elasticmachine committed:
Merge branch 'main' into heatmap-sort

markov00 commented on x-pack/platform/plugins/shared/lens/public/visualizations/heatmap/toolbar_component/style_settings.test.tsx:
```diff
@@ -83,4 +83,135 @@ describe('heatmap style settings', () => {
     expect(updatedOrientationGroup).not.toBeInTheDocument();
   });
+
+  it('should update xSortPredicate when X-axis sort order is changed to descending (string column)', async () => {
```
would be great to add a functional test too that changes the various predicates and checks the charts data sorting order to avoid any regression

markov00 reviewed (commented):
Thanks so much for this work, I've added a couple of comments but looks great otherwise

Commit:
factor out common code

Commit:
Add functional tests for heatmap sort predicates

Add comprehensive functional tests that verify sort predicate changes
and data sorting order to prevent regression. Tests cover X/Y axis
sorting for string and numeric columns with all predicate types.

Commit:
Fix test: re-render component between X and Y axis predicate changes

The test was failing because it made two successive selections without
re-rendering the component. The second onChange handler closed over the
original state prop, missing the first selection's update. Fixed by
tracking state and re-rendering after each setState call.

graphaelli commented on x-pack/platform/plugins/shared/lens/public/visualizations/heatmap/toolbar_component/style_settings.tsx:
```diff
@@ -110,16 +180,100 @@ export function HeatmapVerticalAxisSettings({
+  // Convert current predicate to display value for the select
+  const getDisplayValue = (predicate: string | undefined): string => {
+    if (!predicate) return '';
+    if (predicate === 'dataIndex') return 'dataIndex';
```
good call, done

Commit:
Merge branch 'main' into heatmap-sort

Commit:
Merge branch 'main' into heatmap-sort

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/371532)
* Commit: a6d0bfc725c1ffbadba0028df45ffea3c2f5f330

### Failed CI Steps
* [FTR Configs #3](https://buildkite.com/elastic/kibana-pull-request/builds/371532#019b0b80-ffde-46b3-a698-40c4beb74dbb)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/371532#019b0b80-ffde-46b3-a698-40c4beb74dbb) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/371532/jobs/019b0b80-ffde-46b3-a698-40c4beb74dbb/artifacts/019b0ba5-8982-4dc6-b14d-0e46fe85ef8a) FTR Configs #<span></span>3 / saved objects management apis find with kibana index - relationships `hasReference` and `hasReferenceOperator` parameters search for a reference

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/5f15dc15d6e575073c940c4437fda2b20b9fb023) | [after](https://github.com/elastic/kibana/commit/a6d0bfc725c1ffbadba0028df45ffea3c2f5f330) | diff |
| --- | --- | --- | --- |
| `expressionHeatmap` | 111 | 119 | +8 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/5f15dc15d6e575073c940c4437fda2b20b9fb023) | [after](https://github.com/elastic/kibana/commit/a6d0bfc725c1ffbadba0028df45ffea3c2f5f330) | diff |
| --- | --- | --- | --- |
| `expressionHeatmap` | 16.9KB | 17.0KB | +118.0B |
| `lens` | 1.6MB | 1.6MB | +2.2KB |
| total |  |  | +2.3KB |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/5f15dc15d6e575073c940c4437fda2b20b9fb023) | [after](https://github.com/elastic/kibana/commit/a6d0bfc725c1ffbadba0028df45ffea3c2f5f330) | diff |
| --- | --- | --- | --- |
| `expressionHeatmap` | 16.2KB | 16.6KB | +474.0B |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/5f15dc15d6e575073c940c4437fda2b20b9fb023) | [after](https://github.com/elastic/kibana/commit/a6d0bfc725c1ffbadba0028df45ffea3c2f5f330) | diff |
| --- | --- | --- | --- |
| `expressionHeatmap` | 115 | 123 | +8 |


</details>


### History
* :green_heart: [Build #368521](https://buildkite.com/elastic/kibana-pull-request/builds/368521) succeeded 264050c8df64cbb397ee75295ad13854ca65d5d6

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/371532","number":371532,"commit":"a6d0bfc725c1ffbadba0028df45ffea3c2f5f330"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368521","number":368521,"commit":"264050c8df64cbb397ee75295ad13854ca65d5d6"}],"number":371532}
buildkite-pr-comment-->

markov00 commented on x-pack/platform/plugins/shared/lens/public/visualizations/heatmap/toolbar_component/style_settings.tsx:
```diff
@@ -110,13 +180,61 @@ export function HeatmapVerticalAxisSettings({
+          defaultMessage: 'Sort order',
+        })}
+        fullWidth
+      >
```
We can probably extract this into a component that can be reused in both x and Y

markov00 commented on x-pack/platform/plugins/shared/lens/public/visualizations/heatmap/toolbar_component/style_settings.tsx:
```diff
@@ -20,6 +21,74 @@ import {
+  {
+    value: 'dataIndex',
+    text: i18n.translate('xpack.lens.heatmap.sortOrder.original', {
+      defaultMessage: 'Original order',
```
I would call it "Data order"

graphaelli committed:
reuse predefined options

Applying suggestions from code review

Co-authored-by: Marco Vettorello <vettorello.marco@gmail.com>

graphaelli committed:
Reword "original" order

Commit:
Merge branch 'main' into heatmap-sort

kibanamachine committed:
Changes from node scripts/eslint_all_files --no-cache --fix

graphaelli commented on x-pack/platform/plugins/shared/lens/public/visualizations/heatmap/toolbar_component/style_settings.tsx:
```diff
@@ -20,6 +21,74 @@ import {
+  {
+    value: '',
+    text: i18n.translate('xpack.lens.heatmap.sortOrder.auto', {
+      defaultMessage: 'Auto',
```
@markov00 can I had this off to you at this point? I'm fine to close it if you want to remake as you see fit, I just want to see the capability land in kibana soon

markov00 committed:
Merge branch 'main' into heatmap-sort

markov00 committed:
refactored

markov00 committed:
fix test typo

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/388134)
* Commit: 7a9227576d62b5b3334308e6487effeaca997ecd

### Failed CI Steps
* [Jest Integration Tests #4](https://buildkite.com/elastic/kibana-pull-request/builds/388134#019c0fe7-8458-44a7-a7ec-c95d9ba4b0a8)
* [Jest Integration Tests #4](https://buildkite.com/elastic/kibana-pull-request/builds/388134#019c1014-53f8-4aa2-9e52-3a2101505b26)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/388134#019c0fe7-8458-44a7-a7ec-c95d9ba4b0a8) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/388134/jobs/019c0fe7-8458-44a7-a7ec-c95d9ba4b0a8/artifacts/019c1014-3196-4270-9059-a712da173845) Jest Integration Tests #<span></span>4 / Migration actions - serverless environment closePit closes PointInTime
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/388134#019c1014-53f8-4aa2-9e52-3a2101505b26) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/388134/jobs/019c1014-53f8-4aa2-9e52-3a2101505b26/artifacts/019c1040-0898-4ef5-8af3-8bd73551280e) Jest Integration Tests #<span></span>4 / Migration actions - serverless environment closePit closes PointInTime

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/55be4b139f0e0076c896c1d8a6a8011aae76afc2) | [after](https://github.com/elastic/kibana/commit/7a9227576d62b5b3334308e6487effeaca997ecd) | diff |
| --- | --- | --- | --- |
| `lens` | 1701 | 1702 | +1 |

#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/55be4b139f0e0076c896c1d8a6a8011aae76afc2) | [after](https://github.com/elastic/kibana/commit/7a9227576d62b5b3334308e6487effeaca997ecd) | diff |
| --- | --- | --- | --- |
| `expressionHeatmap` | 111 | 119 | +8 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/55be4b139f0e0076c896c1d8a6a8011aae76afc2) | [after](https://github.com/elastic/kibana/commit/7a9227576d62b5b3334308e6487effeaca997ecd) | diff |
| --- | --- | --- | --- |
| `expressionHeatmap` | 16.9KB | 16.9KB | -68.0B |
| `lens` | 2.0MB | 2.0MB | +1.2KB |
| total |  |  | +1.2KB |

#### Public APIs missing exports
> Total count of every type that is part of your API that should be exported but is not. This will cause broken links in the API documentation system. Target amount is 0. `Run node scripts/build_api_docs --plugin [yourplugin] --stats exports` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/55be4b139f0e0076c896c1d8a6a8011aae76afc2) | [after](https://github.com/elastic/kibana/commit/7a9227576d62b5b3334308e6487effeaca997ecd) | diff |
| --- | --- | --- | --- |
| `expressionHeatmap` | 2 | 3 | +1 |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/55be4b139f0e0076c896c1d8a6a8011aae76afc2) | [after](https://github.com/elastic/kibana/commit/7a9227576d62b5b3334308e6487effeaca997ecd) | diff |
| --- | --- | --- | --- |
| `expressionHeatmap` | 16.2KB | 16.6KB | +410.0B |
| `lens` | 64.2KB | 64.1KB | -19.0B |
| total |  |  | +391.0B |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/55be4b139f0e0076c896c1d8a6a8011aae76afc2) | [after](https://github.com/elastic/kibana/commit/7a9227576d62b5b3334308e6487effeaca997ecd) | diff |
| --- | --- | --- | --- |
| `expressionHeatmap` | 115 | 123 | +8 |


</details>


### History
* :broken_heart: [Build #385786](https://buildkite.com/elastic/kibana-pull-request/builds/385786) failed ba73279b39e2016af515489512753724c830e209
* :broken_heart: [Build #374152](https://buildkite.com/elastic/kibana-pull-request/builds/374152) failed e39a07262375790fb76732bd75e8d3f49a7da688
* :broken_heart: [Build #374141](https://buildkite.com/elastic/kibana-pull-request/builds/374141) failed e94fb596c02ab3418ea072cac4e4203463cb3fb0
* :yellow_heart: [Build #371532](https://buildkite.com/elastic/kibana-pull-request/builds/371532) was flaky a6d0bfc725c1ffbadba0028df45ffea3c2f5f330
* :green_heart: [Build #368521](https://buildkite.com/elastic/kibana-pull-request/builds/368521) succeeded 264050c8df64cbb397ee75295ad13854ca65d5d6

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/388134","number":388134,"commit":"7a9227576d62b5b3334308e6487effeaca997ecd"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/385786","number":385786,"commit":"ba73279b39e2016af515489512753724c830e209"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/374152","number":374152,"commit":"e39a07262375790fb76732bd75e8d3f49a7da688"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/374141","number":374141,"commit":"e94fb596c02ab3418ea072cac4e4203463cb3fb0"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/371532","number":371532,"commit":"a6d0bfc725c1ffbadba0028df45ffea3c2f5f330"}],"number":388134}
buildkite-pr-comment-->

markov00 committed:
Merge branch 'main' into heatmap-sort

markov00 committed:
Merge branch 'main' into heatmap-sort

markov00 committed:
Merge branch 'main' into heatmap-sort

markov00 committed:
Merge branch 'main' into heatmap-sort

andrimal reviewed (commented):
@markov00 Tested locally and seems working nicely! 

I found only one inconsistency between ESQL and DSL which could definitely be part of the follow up task: https://github.com/elastic/elastic-charts/issues/2782

- For ESQL charts, a date axis is sorted with alphabetical order.

https://github.com/user-attachments/assets/e5fc9e54-13dd-4d89-97c1-befff65b8867

- But a DSL date axis doesn't, which feels a bit "buggy":

https://github.com/user-attachments/assets/68935ba0-9f11-4464-bdee-d583c87c6aee

I think the order is forced here:

https://github.com/elastic/kibana/blob/0509d5ae8781d9d3484cdfc73281eafea33edfb7/src/platform/plugins/shared/chart_expressions/expression_heatmap/public/components/heatmap_component.tsx#L442-L462

Here we are using time scale and chronological ascending order.

If we were to add a check for this to run only when  `xSortPredicate === 'none'` we would achieve explicit asc/desc sort but we loose the nicely formatted interval labels:

https://github.com/user-attachments/assets/8518ce95-8c6f-4912-9741-f3a68032393b

markov00 committed:
Merge branch 'main' into heatmap-sort

markov00 committed:
disable sorting for timeseries heatmap

markov00 committed:
add comment to ESQL search expression where the the interval param is missing

markov00 committed:
Add support for sorting predicate to Lens API

markov00 committed:
Merge branch 'main' into heatmap-sort

andrimal reviewed (commented):
Retested the timeseries case and the sorting option is nicely disabled 🎉 

While testing, though I came across the Other bucket sorting logic as shown in the recording. Should we maybe exclude the Other bucket from the sorting logic to always appear as last? What do you think?


https://github.com/user-attachments/assets/647e67e5-484f-473f-95e8-3bcf7f980c11

andrimal reviewed (approved):
Approving to unblock! Tested locally and found no issues ✨ 

The 'Other' bucket will be handled in a follow up!
